sda-commons-client-oidc/src/main/java/org/sdase/commons/client/oidc/OidcClient.java

```diff
@@ -0,0 +1,132 @@
+package org.sdase.commons.client.oidc;
+
+import static org.sdase.commons.client.jersey.proxy.ApiClientInvocationHandler.createProxy;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.concurrent.TimeUnit;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import org.glassfish.jersey.client.proxy.WebResourceFactory;
+import org.sdase.commons.client.jersey.ClientFactory;
+import org.sdase.commons.client.oidc.model.*;
+import org.sdase.commons.client.oidc.rest.model.OpenIdDiscoveryResource;
+import org.sdase.commons.client.oidc.rest.model.TokenResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Service that can retrieve a new access token. */
+public class OidcClient {
+
+  private static final String CACHE_KEY = "oidcResultKey";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OidcClient.class);
+
+  private final Client client;
+
+  private final OidcConfiguration config;
+
+  private final OpenIdDiscoveryApi discoveryApi;
+
+  private Cache<String, TokenResource> cache;
+
+  public OidcClient(ClientFactory clientFactory, OidcConfiguration config) {
+    this.config = config;
+    if (config.isDisabled()) {
+      LOGGER.warn("OIDC was disabled.");
+      client = null;
+      discoveryApi = null;
+      return;
+    }
+
+    this.client =
+        clientFactory.externalClient(config.getHttpClient()).buildGenericClient("oidc-client");
+    this.discoveryApi =
+        createProxy(
+            OpenIdDiscoveryApi.class,
+            WebResourceFactory.newResource(
+                OpenIdDiscoveryApi.class, client.target(config.getIssuerUrl())));
+  }
+
+  /**
+   * Retrieves a new access token from the token endpoint using the given {@link OidcConfiguration}.
+   */
+  public OidcResult createAccessToken() {
+    if (config.isDisabled()) {
+      LOGGER.debug("OIDC client is disabled");
+      return new OidcResult().setState(OidcState.SKIPPED);
+    }
+
+    TokenResource tokenResource;
+
+    if (cache == null) {
```

Comment:
I think the creation of the cache must be synchronized.

Reply:
I suggested yesterday via Slack to use a [custom eviction policy](https://www.baeldung.com/java-caching-caffeine#2-time-based-eviction) to evict each token after the actual expiration time (maybe a few seconds earlier). If thats implemented, the cache could be initialized in advance.

Reply:
But according the synchronization: Fetching new access tokens could be synchronized as well. As the client is always getting a token for the same user, a second invocation can wait until a concurrent invocation is complete and then use the cache.